# Codes used for evaluation of DORGE prediction

## Codes to process evaluation data can be found at:

https://figshare.com/articles/online_resource/Evaluation_data_processing/12693884


## Dependency

Java v1.8.0, bedops v2.4.39, bedtools v2.29.2

Java SDK: Sun Java SE Development Kit 7 or higher (Links of JDK 8 https://www.oracle.com/java/technologies/javase/javase-jdk8-downloads.html) is needed. Setting JAVA_HOME may also be needed to launch Java at any folders(see http://www.sajeconsultants.com/how-to-set-java_home-on-mac-os-x/?utm_source=rss&utm_medium=rss&utm_campaign=how-to-set-java_home-on-mac-os-x).

As a convenience way, Bioconda users may also try 'conda install -c bioconda java-jdk' to install Java easily.

## Contents

1a. Exon BED file generation
```
cd PhyloP
unzip gencode.v19.annotation.gtf.zip
unzip gencode.v25lift37.annotation.gtf.zip
javac phyloP_Gene_CDS_gtf_generation.java
java phyloP_Gene_CDS_gtf_generation
```
  Input: gencode.v19.annotation.gtf #(Gencode v19 GTF annotation)

  Output: compiled_exons_gene.bed #(Exon BED file)

1b. Generate phyloP46-gene association table (Optional; time-comsuming). Output: candidate_gene_exon_phyloP46.txt
```
#see https://www.biostars.org/p/150152/
wget -r -nc -A.wigFix.gz -nd -l1 -R "index.html*" -e robots=off http://hgdownload.cse.ucsc.edu/goldenPath/hg19/phyloP46way/vertebrate/
for fn in `ls *.wigFix.gz`; do gunzip -c ${fn} | wig2bed - > ${fn}.bed; done
bedops --everything chr*.bed > vertebrate.phyloP46.bed
bedtools sort -i compiled_exons_gene.bed > compiled_exons_gene.sorted.bed
bedmap --faster --mean --echo compiled_exons_gene.sorted.bed vertebrate.phyloP46.bed > candidate_gene_exon_phyloP46.txt
```

1c. Process phyloP46 conservation scores
```
javac phyloP_processing.java
java phyloP_processing
cd ..
```
  Input: candidate_gene_exon_phyloP46.txt

  Output: phyloP46_processed.txt #(phyloP scores for genome-wide genes; Also place at Figure_3/data/ in the DORGE_Codes project)


2a. BioGRID PPI gene-gene interaction data generation and node annotation generation.
```
cd BIOGRID
unzip BIOGRID-ALL-3.5.183.tab2.zip
javac BioGRID_network_metrics_calculation.java
java BioGRID_network_metrics_calculation

javac BioGRID_network_node_anno.java
java BioGRID_network_node_anno

```
  Input: Gene_set_new.txt #(Gene annotation information used by this study)

  Input: Cosmic2Genename.txt #(Mapping file from COSMIC Gene ID to gene symbol)
  
  Input: ensembl_synonymous_to_symbol.txt #(Gene alias conversion to HGNC gene names)

  Input: BIOGRID-ALL-3.5.183.tab2.txt #(Gene name and alias mapping file)
  
  Input: DORGE_prediction.txt #(DORGE prediction results)

  Output: BioGRID_edges_all.txt #(Processed PPI data; Input to Cytoscape)
  
  Output: nodes_anno.txt #(Node annotation file; Input to Cytoscape)


2b. Network metrics and PPI landscape generated by Cytoscape.

  Input: BioGRID_edges_all.txt #(From last step)
  
  Input: nodes_anno.txt #(From last step)

  Output: BioGRID_network_node_metrics.csv

  Internal: BioGRID_Cytoscape_session.cys #(BioGRID network Cytoscape session; Also place at Figure_4/data/BioGRID/ in the DORGE_Codes project)

  Output: BioGRID PPI landscape.png #(BioGRID landscape picture; Place at Figure_4/Raw_figures/ in the DORGE_Codes project and rename as Figure_4A_BioGRID_network_landscape.png)


2c. BioGRID PPI Network metrics processing.
```
javac BioGRID_network_metrics_mapping_to_genes.java
java BioGRID_network_metrics_mapping_to_genes
cd ..
```
  Input: BioGRID_network_node_metrics.csv #(From last step)

  Input: Gene_set_new.txt #(Gene annotation information used by this study)

  Input: ensembl_synonymous_to_symbol.txt #(Gene alias conversion to HGNC gene names)

  Output: BioGRID_network_node_metrics_processed.txt #(Including Gene_betweenness_centrality, Gene_closeness_centrality and Gene_Degree)


3a. PharmacoDB data processing.
```
cd PharmacoDB
javac PharmacoDB_network_data_processing.java
java PharmacoDB_network_data_processing
cd ..
```

  Input: gene_drugs.csv (Downloaded from PharmacoDB)

  Input: Gene_set_new.txt #(Gene annotation information used by this study)

  Input: ensembl_synonymous_to_symbol.txt #(Gene alias conversion to HGNC gene names)
  
  Input: genes_201905141625.csv #(Gene annotation information downloaded from PharmacoDB)
  
  Input: drugs_201905141624.txt #(Drug/compound annotation information downloaded from PharmacoDB)
  
  Input: Cosmic2Genename.txt #(Mapping file from COSMIC Gene ID to gene symbol)
  
  Input: DORGE_prediction.txt #(DORGE prediction results)

  Output: drug_gene_nr.txt #(All Gene-drug/compound interaction)

  Output: gene_degree_pharmacodb.txt #(Gene degree metrics)
  
  Output: nodes_anno.txt #(Gene and drug/compound annotation file used by Cytoscape)

  Output: edges_all.txt #(Sampled gene-drug/compound interaction used by Cytoscape)
  
  
3b. PharmacoDB network landscape generated by Cytoscape.

  Input: edges_all.txt (From last step)
  
  Input: nodes_anno.txt (From last step)

  Internal: PharmacoDB_Cytoscape_session.cys #(PharmacoDB network Cytoscape session; Also place at Figure_4/data/PharmacoDB/)
  
  Output: BioGRID_network_node_metrics.csv

  Output: PharmacoDB_landscape_annotated.png #(PharmacoDB Landscape picture; Place at Figure_4/Raw_figures/ in the DORGE_Codes project and rename as Figure_S5D_PharmacoDB_landscape.png)

4. OncoLnc data processing.
```
cd OncoLnc
javac -cp poi-3.14-20160307.jar:poi-ooxml-3.14-20160307.jar:poi-ooxml-schemas-3.14-20160307.jar:xmlbeans-2.6.0.jar oncolnc_data_processing.java
java -cp poi-3.14-20160307.jar:poi-ooxml-3.14-20160307.jar:poi-ooxml-schemas-3.14-20160307.jar:xmlbeans-2.6.0.jar:. oncolnc_data_processing
cd ..
```
  Input: xlsx files in the xlsx folder #(xlsx files downloaded from OncoLnc website)
  
  Input: Top_predicted_TSG_info_without_in_databases.txt #(TSG gene list without TSG database coverage; Data are taken from Tables/Data_file_S1/ folder in the DORGE_Codes project)

  Input: Top_predicted_OG_info_without_in_databases.txt #(OG gene list without OG database coverage; Data are taken from Tables/Data_file_S1/ folder in the DORGE_Codes project)
   
  Input: Gene_set_new.txt #(Gene annotation information used by this study)
  
  Internal: oncolnc_combined.txt #(OncoLnc data in tabular format)
  
  Output: HR_information_for_database_undocumented_novel_TSGs.txt #(HR and related P-value information for TSGs without database coverage; Place at Tables/Data_file_S1 in the DORGE_Codes project)
  
  Output: HR_information_for_database_undocumented_novel_OGs.txt #(HR and related P-value information for OGs without database coverage; Place at Tables/Data_file_S1 in the DORGE_Codes project)
  
  Output: oncolnc_all_genes.txt #(HR and related P-value information for all genes; Place at Figure_3/data in the DORGE_Codes project)
  

5. Drug Gene Budger (DGB) data processing.
```
cd DGB
javac -cp ../OncoLnc/poi-3.14-20160307.jar:../OncoLnc/poi-ooxml-3.14-20160307.jar:../OncoLnc/poi-ooxml-schemas-3.14-20160307.jar:../OncoLnc/xmlbeans-2.6.0.jar DGB_processing.java
java -cp ../OncoLnc/poi-3.14-20160307.jar:../OncoLnc/poi-ooxml-3.14-20160307.jar:../OncoLnc/poi-ooxml-schemas-3.14-20160307.jar:../OncoLnc/xmlbeans-2.6.0.jar:. DGB_processing
cd ..
```

  Input: DORGE_prediction.txt #(DORGE prediction results)
  
  Input: DORGE_predicted_geneID.txt #(Gene symbol to Gene ID mapping file, generated by Metascape)
  
  Input: extract folder including gene-specific excel files downloaded from the DGB database, including all predicted TSG/OGs
  
  Output: DGB_creeds_downregulated_combined.txt #(Flatted file from xlsx files, only downregulated OGs information is kept; Place at Figure_4/data/DGB in the DORGE_Codes project)

  Output: DGB_creeds_upregulated_combined.txt #(Flatted file from xlsx files, only upregulated TSGs information is kept; Place at Figure_4/data/DGB in the DORGE_Codes project)

  Output: DGB_creeds_downregulated_simplified.txt #(Flatted file from xlsx files, only records with downregulated OGs and Q-value < 0.05 are kept; For multiple records in same gene and drug relationship, only the first record is kept; Place at Figure_4/data/DGB in the DORGE_Codes project)
  
  Output: DGB_creeds_upregulated_simplified.txt #(Flatted file from xlsx files, only records with upregulated TSGs and Q-value < 0.05 are kept; For multiple records in same gene and drug relationship, only the first record is kept; Place at Figure_4/data/DGB in the DORGE_Codes project)


6. ATAC-seq data processing.
```
cd ATAC_seq
javac ATACseq_prepare.java
java ATACseq_prepare
cd ..
```

  Input: pan_cancer_ATAC_peaks.bed #(Data S2 of https://science.sciencemag.org/content/362/6413/eaav1898)

  Input: Gene_set_new.txt #(Gene annotation information used by this study)
  
  Input: compiled_genelist_promoter_hg19.bed #(Gene promoters defined as [-500, 1000] around TSSs; From Gene_annotation/Step4_gene_promoter_genebody_BED_generation folder)

  Output: pan_cancer_ATACseq_peaks_processed.txt #(ATAC-seq normalized Peak Scores for genome-wide genes)
  
  
7. Epigenetic regulator gene compilation
```
cd Epigenetic_regulators
javac Epigenetic_regulator_combination.java
java Epigenetic_regulator_combination
cd ..
```

  Input: epigenetic_regulator.csv #(ER genes defined by https://genome.cshlp.org/content/29/4/532)

  Input: EpiFactor_ER.csv #(ER genes in EpiFactor database)

  Output: ER_genes.txt #(Compiled ER gene list; Place at Figure_3/data/ in the DORGE_Codes project)
